coreax/solvers/coresubset.py

```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
             self.unroll,
         )
         return refined_coreset, solver_state
+
+
+def _greedy_kernel_inducing_points_loss(
+    candidate_coresets: Array,
+    responses: Array,
+    feature_gramian: Array,
+    regularisation_parameter: float,
+    identity: Array,
+    least_squares_solver: RegularisedLeastSquaresSolver,
+) -> Array:
+    """
+    Given an array of candidate coreset indices, compute the greedy KIP loss for each.
+
+    Primarily intended for use with :class`GreedyKernelInducingPoints.
+
+    :param candidate_coresets: Array of indices representing all possible "next"
+        coresets
+    :param responses: Array of responses
+    :param feature_gramian: Feature kernel gramian
+    :param regularisation_parameter: Regularisation parameter for stable inversion of
+        array, negative values will be converted to positive
+    :param identity: identity matrix
+    :param least_squares_solver: Instance of
+        :class:`coreax.least_squares.RegularisedLeastSquaresSolver`
+
+    :return: GreedyKernelInducingPoints loss for each candidate coreset
+    """
+    # Extract all the possible "next" coreset feature gramians, cross feature gramians
+    # and coreset response vectors.
+    coreset_gramians = feature_gramian[
+        (candidate_coresets[:, :, None], candidate_coresets[:, None, :])
+    ]
+    coreset_cross_gramians = feature_gramian[candidate_coresets, :-1]
+    coreset_responses = responses[candidate_coresets]
+
+    # Solve for the kernel ridge regression coefficients for each possible coreset
+    coefficients = least_squares_solver.solve_stack(
+        arrays=coreset_gramians,
+        regularisation_parameter=regularisation_parameter,
+        targets=coreset_responses,
+        identity=identity,
+    )
+
+    # Compute the loss function, making sure that we remove the padding on the responses
+    predictions = (coreset_cross_gramians * coefficients).sum(axis=1)
+    loss = ((predictions - 2 * responses[:-1, 0]) * predictions).sum(axis=1)
+    return loss
+
+
+# pylint: disable=too-many-locals
+class GreedyKernelInducingPoints(
+    RefinementSolver[_SupervisedData, GreedyKernelInducingPointsState],
+    ExplicitSizeSolver,
+):
+    r"""
+    Apply GreedyKernelInducingPoints to a supervised dataset.
```

Comment:
"GreedyKernelInducingPoints" should either be in `backticks` or have spaces ("Greedy Kernel Inducing Points"). Same comment for the other instances of it in this docstring.